contracts/crosschain/ERC7786Aggregator.sol

```diff
@@ -0,0 +1,278 @@
+                emit ExecutionFailed(id);
+            }
+        }
+
+        return IERC7786Receiver.executeMessage.selector;
+    }
+
+    // =================================================== Getters ===================================================
+
+    function getGateways() public view virtual returns (address[] memory) {
+        return _gateways.values();
+    }
+
+    function getThreshold() public view virtual returns (uint8) {
+        return _threshold;
+    }
+
+    function getRemoteRouter(string calldata caip2) public view virtual returns (string memory) {
+        string memory router = _remoteRouters[caip2];
+        if (bytes(router).length == 0) revert ERC7786AggregatorRemoteNotRegistered(caip2);
+        return router;
+    }
+
+    // =================================================== Setters ===================================================
+
+    function addGateway(address gateway) public virtual onlyOwner {
+        _addGateway(gateway);
+    }
+
+    function removeGateway(address gateway) public virtual onlyOwner {
+        _removeGateway(gateway);
+    }
+
+    function setThreshold(uint8 newThreshold) public virtual onlyOwner {
+        _setThreshold(newThreshold);
+    }
+
+    function registerRemoteRouter(string memory caip2, string memory router) public virtual onlyOwner {
+        _registerRemoteRouter(caip2, router);
+    }
+
+    function pause() public virtual onlyOwner {
+        _pause();
+    }
+
+    function unpause() public virtual onlyOwner {
+        _unpause();
+    }
+
+    /// @dev Recovery method in case ether is ever received through {executeMessage}
+    function drainEth(address payable to) public virtual onlyOwner {
+        Address.sendValue(to, address(this).balance);
+    }
+
+    // ================================================== Internal ===================================================
+
+    function _addGateway(address gateway) internal virtual {
+        if (!_gateways.add(gateway)) revert ERC7786AggregatorGatewayAlreadyRegistered(gateway);
+        emit GatewayAdded(gateway);
+    }
+
+    function _removeGateway(address gateway) internal virtual {
+        if (!_gateways.remove(gateway)) revert ERC7786AggregatorGatewayNotRegistered(gateway);
+        if (_threshold > _gateways.length()) revert ERC7786AggregatorThresholdViolation();
+        emit GatewayRemoved(gateway);
+    }
+
+    function _setThreshold(uint8 newThreshold) internal virtual {
+        if (newThreshold == 0 || _threshold > _gateways.length()) revert ERC7786AggregatorThresholdViolation();
+        _threshold = newThreshold;
+        emit ThresholdUpdated(newThreshold);
+    }
+
+    // NOTE: once a router is registered for a given chainId, it cannot be updated
+    function _registerRemoteRouter(string memory caip2, string memory router) internal virtual {
+        if (bytes(_remoteRouters[caip2]).length > 0) revert ERC7786AggregatorRemoteAlreadyRegistered(caip2);
```

Comment:
Is there a concrete reason why it shouldn't be possible to overwrite an existing value?

Reply:
The logic was that the admin could setup new chains, but not update/break existing links. The is to ensure that something that worked in the past would not be broken by the admin.

This logic comes from the Axelar adaptor.

But given that the admin can add gateways,  remove gateways, and update the threshold ... maybe that doesn't bring much value to the end user.

Should we remove that check ?

Reply:
Let's remove it.